flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverRowProcessFunction.scala

```diff
@@ -0,0 +1,182 @@
+    
+    counterState = getRuntimeContext.getState(processedCountDescriptor)
+    
+    val smallesTimestampDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("smallesTSState", classOf[Long])
+    
+    smallestTsState = getRuntimeContext.getState(smallesTimestampDescriptor)
+    
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+    
+    val currentTime = ctx.timerService().currentProcessingTime()
+    var i = 0
+
+    accumulators = accumulatorState.value()
+    // initialize state for the first processed element
+    if(accumulators == null){
+      accumulators = new Row(aggregates.length)
+      while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      }
+    }
+    
+    // get smallest timestamp 
+    smallestTs = smallestTsState.value()
+    if(smallestTs == 0L){
+      smallestTs = currentTime
+    }
+    // get previous counter value
+    counter = counterState.value()
+    
+    if (counter == precedingOffset) {
+      val retractTs = smallestTs
+      val retractList = rowMapState.get(smallestTs)
+
+      // get oldest element beyond buffer size   
+      // and if oldest element exist, retract value
+      i = 0
+      while (i < aggregates.length) {
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        aggregates(i).retract(accumulator, retractList.get(0).getField(aggFields(i)))
+        i += 1
+      }
+      retractList.remove(0)
+      counter -= 1
+      // if reference timestamp list not empty, keep the list
+      if (!retractList.isEmpty()) {
+        rowMapState.put(retractTs, retractList)
+      } // if smallest timestamp list is empty, remove and find new smallest
+      else {
+        rowMapState.remove(retractTs)
+        val iter = rowMapState.keys.iterator()
+        var currentTs: Long = 0L
+        var newSmallesTs: Long = Long.MaxValue
+        while(iter.hasNext){
+          currentTs = iter.next
+          if(currentTs < newSmallesTs){
+            newSmallesTs = currentTs
+          }
+        }
+        smallestTs = newSmallesTs
+      }
+    }
+
+    // copy forwarded fields in output row
+    i = 0
+    while (i < forwardedFieldCount) {
+      output.setField(i, input.getField(i))
+      i += 1
+    }
+
+    // accumulate current row and set aggregate in output row
+    i = 0
+    while (i < aggregates.length) {
+      val index = forwardedFieldCount + i
+      val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+      aggregates(i).accumulate(accumulator, input.getField(aggFields(i)))
+      output.setField(index, aggregates(i).getValue(accumulator))
+      i += 1
+    }
+
+    // update map state, accumulator state, counter and timestamp
+    if (rowMapState.contains(currentTime)) {
+      rowMapState.get(currentTime).add(input)
```

Comment:
This will not modify the state but just the list object returned from the state. We need to explicitly `put` the updated list into the `MapState`. It works for in-memory state backends because all objects are hold on the heap.